File: python/leetcode/210.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        from collections import deque
        graph = { i:set() for i in range(numCourses)} # v:income number
        neigh = [set() for i in range(numCourses)] # graph
        res, cnts = [], 0
        for [x, y] in prerequisites:
            graph[x].add(y)
            neigh[y].add(x) 
        queue = deque([i for i in graph if not graph[i]])
        while queue:
            v = queue.popleft()
            res.append(v)
            cnts += 1
            for node in neigh[v]:
                graph[node].remove(v)
                if not graph[node]: queue.append(node)
        return res if cnts == numCourses else []

# DFS
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        graph = [set() for i in range(numCourses)]
        visit = [0 for i in range(numCourses)]
        res = []
        for x, y in prerequisites:
            graph[y].add(x)
        for i in range(numCourses):
            if not self.dfs(i, graph, visit, res): return []
        return res[::-1]
    
    def dfs(self, i, graph, visit, res):
        if visit[i] == 1: return True
        if visit[i] == -1: return False
        visit[i] = -1
        for j in graph[i]:
            if not self.dfs(j, graph, visit, res): return False
        visit[i] = 1
        res.append(i)
        return True

File: cpp/leetcode/210.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<unordered_set<int>> graph(numCourses);
        vector<int> degrees(numCourses, 0);
        queue<int> zeros;
        vector<int> res;
        for(auto pre: prerequisites) graph[pre.second].insert(pre.first);
        for (auto v : graph)
            for (int neigh : v) degrees[neigh]++
        for (int i = 0; i < numCourses; i++)
            if (!degrees[i]) zeros.push(i);
        for (int i = 0; i < numCourses; i++) {
            if (zeros.empty()) return {};
            int node = zeros.front();
            zeros.pop();
            res.push_back(node);
            for (int neigh : graph[node]) {
                if (!--degrees[neigh]) zeros.push(neigh);
            }
        }
        return res;
    }
};



//DFS
class Solution {
public:
    vector<int> findOrder(int num, vector<pair<int, int>>& pres) {
        unordered_map<int, unordered_set<int>> graph;
        vector<int> visit(num, 0);
        vector<int> res;
        for(auto pre : pres){
            graph[pre.second].insert(pre.first);
        }
        for(int i = 0; i < num; i++){
            if(!dfs(i, graph, visit, res)) return {};
        }
        reverse(res.begin(), res.end());
        return res;
    }
    
    bool dfs(int i, unordered_map<int, unordered_set<int>> &graph, vector<int> &visit, vector<int> &res){
        if(visit[i] == 1) return true;
        if(visit[i] == -1) return false;
        visit[i] = -1;
        for(auto j : graph[i]){
            if(!dfs(j, graph, visit, res)) return false;
        }
        res.push_back(i);
        visit[i] = 1;
        return true;
    }
};